### PR TITLE
fix: auto-replace pinning

### LINKED ISSUE #20625: docker digestPin is no longer working for multiple images in same file
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### If you're self-hosting Renovate, select which platform you are using.

None

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

I noticed today that when using docker:digestPin, renovate is no longer updating all image/from statements within the dockerfile.  It would appear that it is only updating the last dependency, although the PR body table contains all of the dependencies and updates

Minimal reproduction
- repository - https://github.com/setchy/renovate-docker-pinning-issue
- pin PR with the two dependencies correctly identified in issue body - https://github.com/setchy/renovate-docker-pinning-issue/pull/5
- only the last dependency was pinned - https://github.com/setchy/renovate-docker-pinning-issue/pull/5/files

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by setchy**
## Changes

revert to previous confirmIfDepUpdated logic, with new conditional to check if the updateType is replacement

## Context

- Fixes https://github.com/renovatebot/renovate/issues/20625

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/update/branch/auto-replace.ts [hunk 1]
```diff
@@ -179,6 +177,8 @@ export async function doAutoReplace(
     const depIndex = existingContent.indexOf(depName!);
     const valIndex = existingContent.indexOf(currentValue!);
     searchIndex = depIndex < valIndex ? depIndex : valIndex;
+    // } else if (upgrade.replacementPrefix) {
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.ts:180]: ?

**Comment by rarkins** [comment]
Was there a particular PR which broke it recently?

File: lib/workers/repository/update/branch/auto-replace.spec.ts [hunk 2]
```diff
@@ -1044,12 +1044,12 @@ describe('workers/repository/update/branch/auto-replace', () => {
         FROM ubuntu:18.04
       `;
       upgrade.manager = 'dockerfile';
+      upgrade.updateType = 'replacement';
+      upgrade.pinDigests = true;
       upgrade.depName = 'ubuntu';
       upgrade.currentValue = '18.04';
       upgrade.currentDigest = undefined;
       upgrade.depIndex = 0;
-      upgrade.pinDigests = true;
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:1051]: for consistency. 

made it easier to review the recently added tests from https://github.com/renovatebot/renovate/pull/20317

in doing so i found several that didn't have updateType=replacement, and [this test ](https://github.com/renovatebot/renovate/pull/20627/files/80630f02a00470e4404eb487ba8ed71072aa2c58#diff-53c104653fd25b77cb8954223f0037f712d0457ed23404fe2d465d99cf97e28fR1136)which shouldn't have had pinDigests=true as it was a [duplicating this](https://github.com/renovatebot/renovate/pull/20627/files/80630f02a00470e4404eb487ba8ed71072aa2c58#diff-53c104653fd25b77cb8954223f0037f712d0457ed23404fe2d465d99cf97e28fR1249)

File: lib/workers/repository/update/branch/auto-replace.ts (hunk 1 above)
Comment (setchy) [reply lib/workers/repository/update/branch/auto-replace.ts]: 🙈 sorry - not sure how this got onto this branch - 🧹 removed

**Comment by setchy** [comment]
> Was there a particular PR which broke it recently?

yes, seems like https://github.com/renovatebot/renovate/pull/20317 did

**Comment by setchy** [comment]
Interestingly, I just re-tested the main branch a real repository but commented out [the !firstUpdate confirmIfDepUpdated](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/branch/auto-replace.ts#L222) (like i had to in the [open PR for replacement name-only](https://github.com/renovatebot/renovate/pull/20629)).  This change was able to resolve pinning as well.

That leads me to believe that the issue to resolve is more-so in how the [!firstUpdate confirmIfDepUpdated](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/branch/auto-replace.ts#L222) logic behaves

**Comment by setchy** [comment]
I've rewritten `confirmIfDepUpdated` so that each check is more explicit and avoids flipping between returning false or true.

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
needs coverage fixes 😉

**Comment by setchy** [comment]
> needs coverage fixes 😉

test-case added.  appreciate some extra eyes on it - felt like a weird scenario to me

**Review by viceice — COMMENTED** [review]
🤷‍♂️

**Review by viceice — DISMISSED** [review]
🙈

File: lib/workers/repository/update/branch/auto-replace.ts [hunk 3]
```diff
@@ -63,35 +55,56 @@ export async function confirmIfDepUpdated(
...
-    } else if (newDigest) {
-      return true;
+  if (upgrade.newDigest) {
+    if (
+      (upgrade.pinDigests &&
+        !upgrade.currentDigest &&
+        upgrade.updateType === 'pinDigest') ||
+      upgrade.currentDigest
+    ) {
+      if (upgrade.newDigest !== newUpgrade.currentDigest) {
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.ts:91]: can we combine to one condition? i only see 3 nested if and no else.

Comment (setchy) [reply lib/workers/repository/update/branch/auto-replace.ts:91]: I can do if we feel that's more legible.  I left them split out initially as I found it much easier to comprehend to different scenarios being tested for

Comment (viceice) [reply lib/workers/repository/update/branch/auto-replace.ts:91]: I'd like to have less nested code, which is easier  to read for me

Comment (setchy) [reply lib/workers/repository/update/branch/auto-replace.ts:91]: will update the PR shortly

**Review by viceice — COMMENTED** [review]
missing coverage

**Comment by setchy** [comment]
@viceice - should be all now 🤞

File: lib/workers/repository/update/branch/auto-replace.ts [hunk 4]
```diff
@@ -63,35 +55,52 @@ export async function confirmIfDepUpdated(
...
-  if (!currentDigest) {
-    if (!pinDigests) {
-      return true;
-    } else if (newDigest) {
-      return true;
+  if (
+    upgrade.updateType !== 'replacement' &&
+    (upgrade.pinDigests || upgrade.currentDigest)
+  ) {
+    if (upgrade.newDigest && upgrade.newDigest !== newUpgrade.currentDigest) {
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.ts:88]: will there be another condition later or why you still left the nested if?

Comment (setchy) [reply lib/workers/repository/update/branch/auto-replace.ts]: Resolved with a further update and aggregation of logic

**Review by viceice — APPROVED** [review]

**Pull request merged**